crates/subspace-node/src/lib.rs

```diff
@@ -156,15 +159,29 @@ impl SubstrateCli for Cli {
     }
 
     fn load_spec(&self, id: &str) -> Result<Box<dyn ChainSpec>, String> {
-        Ok(match id {
-            "testnet" => Box::new(chain_spec::testnet_config_json()?),
-            "testnet-compiled" => Box::new(chain_spec::testnet_config_compiled()?),
-            "dev" => Box::new(chain_spec::dev_config()?),
-            "" | "local" => Box::new(chain_spec::local_config()?),
-            path => Box::new(chain_spec::SubspaceChainSpec::from_json_file(
-                std::path::PathBuf::from(path),
-            )?),
-        })
+        let mut chain_spec = match id {
+            "testnet" => chain_spec::testnet_config_json()?,
+            "testnet-compiled" => chain_spec::testnet_config_compiled()?,
+            "dev" => chain_spec::dev_config()?,
+            "" | "local" => chain_spec::local_config()?,
+            path => SubspaceChainSpec::from_json_file(std::path::PathBuf::from(path))?,
+        };
+
+        // In case there are bootstrap nodes specified explicitly, ignore those that are in the
+        // chain spec
+        if !self.run.base.network_params.bootnodes.is_empty() {
+            let mut chain_spec_value =
+                serde_json::from_str::<'_, Value>(&chain_spec.as_json(true)?)
+                    .map_err(|error| error.to_string())?;
+            if let Some(boot_nodes) = chain_spec_value.get_mut("bootNodes") {
+                if let Some(boot_nodes) = boot_nodes.as_array_mut() {
+                    boot_nodes.clear();
```

Comment:
This will clear the user provided bootnodes no matter what the network type is? what if I want to run a dev network with multiple nodes, why not clear them just for `testnet-compiled` to minimize the potential side effects if only this kind of network is affected?

Reply:
Maybe it wasn't clear what this does. The default behavior of Substrate is to merge boot nodes from chain spec and CLI arguments into one vector. This change makes CLI arguments replace boot nodes from chain spec, but only when CLI argument is specified. So nothing prevents you to specify the same bootnode as in chain spec + some extra if you want.

Reply:
Okay, only the bootnodes specified in chain_spec get cleared when the CLI bootnodes are not empty. Thanks for clarifying!